File: java/BedrockSlimeFinder.java
public class BedrockSlimeFinder {
    
    private static long umul32_lo(long a, long b) {
        long a_low = a & 0xFFFFL;
        long b_low = b & 0xFFFFL;
        long result = a_low * b_low;
        long carry = result >>> 16;
        
        carry += (a >>> 16) * b_low;
        carry &= 0xFFFFL;
        
        carry += a_low * (b >>> 16);
        
        return ((carry & 0xFFFFL) << 16 | (result & 0xFFFFL)) & 0xFFFFFFFFL;
    }

    public static class MersenneTwister {
        private static final int N = 624;
        private static final int M = 397;
        private static final long MATRIX_A = 0x9908B0DFL;
        private static final long UPPER_MASK = 0x80000000L;
        private static final long LOWER_MASK = 0x7FFFFFFFL;
        
        private long[] mt;
        private int mti;
        
        public MersenneTwister(long seed) {
            mt = new long[624];
            mti = 625;
            initSeed(seed);
        }
        
        private void initSeed(long s) {
            mt[0] = s & 0xFFFFFFFFL;
            for (mti = 1; mti < N; mti++) {
                long t = mt[mti - 1] ^ (mt[mti - 1] >>> 30);
                mt[mti] = (((1812433253L * ((t & 0xFFFF0000L) >>> 16)) << 16) + 
                          (1812433253L * (t & 0xFFFFL)) + mti) & 0xFFFFFFFFL;
            }
        }
        
        public long randomInt() {
            long[] mag01 = {0x0L, MATRIX_A};
            
            if (mti >= N) {
                int kk;
                if (mti == N + 1) initSeed(5489L);

                for (kk = 0; kk < N - M; kk++) {
                    long y = (mt[kk] & UPPER_MASK) | (mt[kk + 1] & LOWER_MASK);
                    mt[kk] = mt[kk + M] ^ (y >>> 1) ^ mag01[(int)(y & 0x1L)];
                }
                for (; kk < N - 1; kk++) {
                    long y = (mt[kk] & UPPER_MASK) | (mt[kk + 1] & LOWER_MASK);
                    mt[kk] = mt[kk + (M - N)] ^ (y >>> 1) ^ mag01[(int)(y & 0x1L)];
                }
                long y = (mt[N - 1] & UPPER_MASK) | (mt[0] & LOWER_MASK);
                mt[N - 1] = mt[M - 1] ^ (y >>> 1) ^ mag01[(int)(y & 0x1L)];
                mti = 0;
            }

            long y = mt[mti++];
            y ^= y >>> 11;
            y ^= (y << 7) & 0x9D2C5680L;
            y ^= (y << 15) & 0xEFC60000L;
            y ^= y >>> 18;
            
            return y & 0xFFFFFFFFL;
        }
    }
    
    public static long getSlimeRandomSeed(int x, int z) {
        long seed = umul32_lo(Integer.toUnsignedLong(x), 522133279L) ^ Integer.toUnsignedLong(z);
        MersenneTwister rng = new MersenneTwister(seed);
        return rng.randomInt();
    }
    
    public static boolean isSlimeChunk(int x, int z) {
        long randomVal = getSlimeRandomSeed(x, z);
        return (randomVal % 10) == 0;
    }
}

File: kotlin/BedrockSlimeFinder.kt
object BedrockSlimeFinder {

    private fun umul32_lo(a: Long, b: Long): Long {
        val a_low = a and 0xFFFFL
        val b_low = b and 0xFFFFL
        val result = a_low * b_low
        var carry = result ushr 16

        carry += (a ushr 16) * b_low
        carry = carry and 0xFFFFL

        carry += a_low * (b ushr 16)

        return ((carry and 0xFFFFL) shl 16 or (result and 0xFFFFL)) and 0xFFFFFFFFL
    }

    class MersenneTwister(seed: Long) {
        companion object {
            private const val N = 624
            private const val M = 397
            private const val MATRIX_A = 0x9908B0DFL
            private const val UPPER_MASK = 0x80000000L
            private const val LOWER_MASK = 0x7FFFFFFFL
        }

        private val mt = LongArray(N)
        private var mti = N + 1

        init {
            initSeed(seed)
        }

        private fun initSeed(s: Long) {
            mt[0] = s and 0xFFFFFFFFL
            for (i in 1 until N) {
                val t = mt[i - 1] xor (mt[i - 1] ushr 30)
                mt[i] = (((1812433253L * ((t and 0xFFFF0000L) ushr 16)) shl 16) +
                        (1812433253L * (t and 0xFFFFL)) + i) and 0xFFFFFFFFL
            }
            mti = N
        }

        fun randomInt(): Long {
            val mag01 = longArrayOf(0L, MATRIX_A)

            if (mti >= N) {
                if (mti == N + 1) initSeed(5489L)

                var kk = 0
                while (kk < N - M) {
                    val y = (mt[kk] and UPPER_MASK) or (mt[kk + 1] and LOWER_MASK)
                    mt[kk] = mt[kk + M] xor (y ushr 1) xor mag01[(y and 0x1L).toInt()]
                    kk++
                }
                while (kk < N - 1) {
                    val y = (mt[kk] and UPPER_MASK) or (mt[kk + 1] and LOWER_MASK)
                    mt[kk] = mt[kk + (M - N)] xor (y ushr 1) xor mag01[(y and 0x1L).toInt()]
                    kk++
                }
                val y = (mt[N - 1] and UPPER_MASK) or (mt[0] and LOWER_MASK)
                mt[N - 1] = mt[M - 1] xor (y ushr 1) xor mag01[(y and 0x1L).toInt()]
                mti = 0
            }

            var y = mt[mti++]
            y = y xor (y ushr 11)
            y = y xor ((y shl 7) and 0x9D2C5680L)
            y = y xor ((y shl 15) and 0xEFC60000L)
            y = y xor (y ushr 18)

            return y and 0xFFFFFFFFL
        }
    }

    fun getSlimeRandomSeed(x: Int, z: Int): Long {
        val seed = umul32_lo(x.toLong() and 0xFFFFFFFFL, 522133279L) xor (z.toLong() and 0xFFFFFFFFL)
        val rng = MersenneTwister(seed)
        return rng.randomInt()
    }

    fun isSlimeChunk(x: Int, z: Int): Boolean {
        val randomVal = getSlimeRandomSeed(x, z)
        return randomVal % 10 == 0L
    }
}